### DISCUSSION TITLE
useContract returning Contract, useContractRead returning undefined

### DISCUSSION THREAD

**Discussion by connorphill** [wevm/wagmi#352] [Q&A]
I am having trouble figuring out what is wrong with my `useContractRead` hook as I mimicked the setup in my `useContract` hook that returned an `ethers.Contract` object, but when I return the promise, it comes back as `undefined`, rather than what I anticipate, which is the call data from the function.

Here is my code setup:

```
import { useProvider, useContract, useContractRead } from 'wagmi'
import ERC721NFT_IPFS from "../../src/artifacts/contracts/ERC721/ERC721NFT_IPFS.sol/ERC721NFT_IPFS.json";

const ERC721Form = ({ props, walletAddress, errorMessage, onSubmit }) => {
    const provider = useProvider(); // Alchemy

    const contractAddress = process.env.ERC721NFT_IPFS_RINKEBY_ADDRESS || "";

    const contract = useContract({
        addressOrName: contractAddress,
        contractInterface: ERC721NFT_IPFS.abi,
        signerOrProvider: provider
    });

    const [{data, error, loading}, read] = useContractRead({
        addressOrName: contractAddress,
        contractInterface: ERC721NFT_IPFS.abi
        },
        "isMintEnabled",
    );
}
```

Where `contract` returns the appropriate Contract:

```
Contract {
  interface: Interface {
    fragments: [
    ...
}
```

and `data, error, loading` returns `undefined, undefined, false`.

In the ABI of my contract I have:
```
{
      "inputs": [],
      "name": "isMintEnabled",
      "outputs": [],
      "stateMutability": "nonpayable",
      "type": "function"
    },
```

which is also present in the Contract so I wouldn't anticipate that part to be the issue. If I drop the contract function reference from the hook there is no change in the data, error and loading output. What could be missing from my hook?

**Comment by brokewhale** [comment]
i am having this same issue, were you able to solve it?

**Reply by connorphill** [reply]
I just updated to `0.3.0`, which has some breaking changes to the `useProvider` setup, but the problem stemmed from the provider not recognizing my Alchemy API and request to the rinkeby network. Since there wasn't an API to interact with the node, the request was not going through with the call. At least that was the change in my setup that resulted in a returned call

**Comment by audsssy** [comment]
Is anyone encountering this issue despite having latest version, e.g., `0.4.12`?

**Reply by brokewhale** [reply]
I was able to solve the issue but really can't remember how. I know the issue came from the way I set up my provider in the config ( inside _app.tsx or any other place you have it ), you should check it out and make sure you are actually using the alchemy provider

**Reply by brokewhale** [reply]
`const { chains, provider, webSocketProvider } = configureChains(
  [chain.polygonMumbai, chain.mainnet, chain.polygon],
  [alchemyProvider({ alchemyId }), publicProvider()]
);`

you can try this, I don't know if this helps
this is for v ^0.4.3, but I am sure there shouldn't be many changes

**Reply by mochi-roe** [reply]
@brokewhale's solution works for me

**Reply by brokewhale** [reply]
Awesome

**Comment by lelerukjaymoh** [comment]
Got the same issue

**Comment by robriks** [comment]
Am having this issue as well but with useContractWrite- it claims to be looking for 'undefined' in the contract abi or ethers interface despite functionName being specified with a function name that definitely exists in the abi

**Reply by jxom** [reply]
A minimal reproducible example on CodeSandbox would be great! If you can do this, I can have a look into the issue for you. :)

**Reply by robriks** [reply]
Thanks for your rapid response! Turns out there was simply one parameter, the name of the function, missing in the documentation example. I'm not sure if I'm just using an outdated version of wagmi but in my project I needed to use this code:

import { useContractWrite } from 'wagmi'

function App() {
  const { data, isError, isLoading, write } = useContractWrite({
    addressOrName: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
    contractInterface: wagmigotchiABI,
    functionName: 'feed',
  },
  'feed' // This string is missing in the documentation for useContractWrite, 
             // so it was trying an undefined function call but code works when this is added
)

  return <button onClick={() => write()}>Feed</button>
}

I have opened a pull request on the file with the example in question though I suspect there may be a reason why it is left out of the documentation (ie changed on recent update?)

**Reply by tmm** [reply]
What version of wagmi are you using? [`useContractWrite`'s API changed](https://wagmi.sh/docs/migration-guide#usecontractwrite=) from <=0.4 to 0.5.

**Reply by robriks** [reply]
It's wagmi version 0.4.5

I can link you my repo of the project with the issue if you like

**Reply by tmm** [reply]
No need. Since you are using `0.4.5`, refer to the `0.4.x` docs at [0.4.x.wagmi.sh](https://0.4.x.wagmi.sh) instead of the `0.5.x` docs (current stable version) at [wagmi.sh](https://wagmi.sh).

**Comment by robriks** [comment]
Thank you!!

**Comment by halalanft** [comment]
This bug is really annoying as in version 0.9.0 I'm still facing the issue.

in `pages/index.js` I put the following code:
```js
    const { data: owner } = useContractRead({
        addressOrName: contractAddres.<ContractName>,
        contractInterface: proudABI.abi,
        functionName: 'owner',
    });

it returned undefined. I put in the component once and it's not a problem. 
    
```

**Reply by tmm** [reply]
Starting in `wagmi@0.7`, `addressOrName` and `contractInterface` were [renamed to `address` and `abi`](https://wagmi.sh/react/migration-guide#addressorname-and-contractinterface-renamed-for-contract-hooks) respectively.

**Reply by halalanft** [reply]
@tmm thank you for your correction.

**Comment by theMacAli** [comment]
You can solve this problem by use OnSuccess function and a useState. 

const [nftData, setNftData] = useState(null);

const { data, isError, isLoading} = useContractRead({
    address: contractAddress,
    abi: contractABI.abi,
    functionName: "fetchMyNfts",
    **onSuccess(data) {
      console.log('Success', data)
      setNftData(data);
    },**
})

**Reply by halalanft** [reply]
Thanks, I use this and it works.

**Reply by ummartoor** [reply]
it's not working for me

**Reply by ummartoor** [reply]
and why are you using like (contractABI.abi) .abi after the name of imported contractABI?

**Reply by ummartoor** [reply]
```
 const { data, isError, isLoading } = useContractRead({
    address: contract.address,
    abi: contract.abi,
    functionName: "symbol",
    onSuccess(data) {
      console.log("Success", data);
      setNftData(data);
    },
  });
```
still getting undefined

**Comment by seyiram** [comment]
I'm also facing a similar problem. I tried your solution but unfortunately didn't work: 
 const [walletInfo, setWalletInfo] = useState(null);
 const { data } = useContractRead({
    address: contractAddress,
    abi: ContractABI.abi,
    functionName: "getWallet",
    onSuccess(data) {
      console.log("success", data);
      setWalletInfo(data);
    },
  });
data is returning undefined

**Reply by theMacAli** [reply]
your data will be in **NftData** and not in data

**Reply by theMacAli** [reply]
in your coding case, it will be in WalletInfo state and not in data

**Reply by seyiram** [reply]
walletInfo is also null. Not sure what I'm doing wrong.

**Reply by topcook** [reply]
I also have same issue. Instead of undefined, it returns null. I am not sure why this error happens. In the previous contracts, which I have written, it shows correct results. But if I use newly developed contract, then it shows null. In some cases(happened in several test cases), it shows 0 or 0x0000..0, "", false, [ ] according to the return parameters.

**Comment by mastmaster** [comment]
How can call useContractRead function of click button instead of load!

**Reply by tmm** [reply]
```tsx
const [enabled, setEnabled] = React.useState(false)
const { data } = useContractRead({
  enabled,
  address: '0x…',
  abi: […],
  functionName: 'foo',
  args: […],
})

<button onClick={() => setEnabled(true)}>Read</button>
```

**Reply by mastmaster** [reply]
Thank you, but this is not an ideal solution for me!

**Comment by ArtGurianov** [comment]
Just got the same issue. For me the solution was to change the Wagmi config setup.
Instead of building up Wagmi client, I had to pass my Viem client as Wagmi client.
Simplified code looks like this:
```
const viemClient = createPublicClient({
  chain: localhost, //imported from viem/chains
  transport: http() //imported from viem
});

export const wagmiConfig = createConfig({
  chains: [mainnet, localhost],
  connectors: [
    walletConnect({ projectId }),
    metaMask()
  ],
  client: () => viemClient,
});

```

**Reply by tmm** [reply]
Passing a Viem client like this is not a good idea as it will not update based on chain changes (e.g. switchChain, connector events).

More info https://wagmi.sh/react/api/createConfig#client
